ui: Custom graph debug page

mrtracy opened the pull request:
Two commits
+ A small update to LineGraph to support graphs without line-related hover events
+ The custom graph debug page

Adds a debug page which allows users to create a "custom" graph,
displaying any combination of metrics they wish. The currently selected
metrics are stored in the URL, so graphs can be shared by copying the
link.

<img width="1431" alt="screen shot 2018-02-25 at 9 22 44 pm" src="https://user-images.githubusercontent.com/6969858/36650698-d58ca892-1a72-11e8-8b1a-56ba7a9ce5dc.png">

Commit:
ui: Custom Graph Debug Page

Adds a debug page which allows users to create a "custom" graph,
displaying any combination of metrics they wish. The currently selected
metrics are stored in the URL, so graphs can be shared by copying the
link.

Commit:
ui: Make linegraph hover options actually optional

Instantiating a graph without these options was causing numerous
rendering issues.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23070)
<!-- Reviewable:end -->

petermattis commented:
Yay!

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-:-L6HY5-aYRKtUTVtSlr4:bb77rlh)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Nice work! A couple small comments.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ui/src/views/reports/containers/customgraph/index.tsx, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-L6Kr-YUIQ0rHllMnI6-:-L6Kr-YUIQ0rHllMnI60:bbhya3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e0ee011a4774b5a6ae8eb9a327ed54338ad28ee/pkg/ui/src/views/reports/containers/customgraph/index.tsx#L194)):*
> ```XML
>       table = (
>         <table>
>           <thead>
> ```

Layout nit: contents of the <td>s for these columns seem to be centered, but the <th>s aren't. Makes it a bit hard to read; centering the <th>s would be nice.

---

*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-L6KqitdJ4yTB0jt6Dui:-L6KqiteuqmcU7uO0aNU:b-77pyvq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e0ee011a4774b5a6ae8eb9a327ed54338ad28ee/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L93)):*
> ```XML
>   derivative?: TimeSeriesQueryDerivative;
>   aggregator?: TimeSeriesQueryAggregator;
>   downsampler?: TimeSeriesQueryAggregator;
> ```

So, the new page uses these new fields, while the old page uses the old one? I see why (was nice to just be able to say `<Metric … nonNegativeRate />` in the JSX, but I'm a little bummed about the repetition here.

I guess changing all the metrics to `<Metric … derivative={NON_NEGATIVE_RATE} />` or whatever would be a lot of work, but might be worth it.

Anyway, maybe at least add a little comment here, explaining the overlapping fields?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-:-L6Kqp_fkd2ET87DcpSH:b-hwcq9e)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Updated the appearance of the table a bit:

![screen shot 2018-02-27 at 4 18 26 pm](https://user-images.githubusercontent.com/6969858/36755388-06ba69d2-1bda-11e8-978f-4e1eef93489c.png)

mrtracy commented:
Review status: 1 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/ui/src/views/reports/containers/customgraph/index.tsx, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-L6Kr-YUIQ0rHllMnI6-:-L6O7t7V6XRgZYWP0fFz:bx84yhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e0ee011a4774b5a6ae8eb9a327ed54338ad28ee/pkg/ui/src/views/reports/containers/customgraph/index.tsx#L194)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Layout nit: contents of the <td>s for these columns seem to be centered, but the <th>s aren't. Makes it a bit hard to read; centering the <th>s would be nice.
</blockquote></details>

Updated the appearance a bit, see screenshot in the PR.

---

*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-L6KqitdJ4yTB0jt6Dui:-L6O7vyQdX7W8hOfN3Dq:b-tdbn1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e0ee011a4774b5a6ae8eb9a327ed54338ad28ee/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L93)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

So, the new page uses these new fields, while the old page uses the old one? I see why (was nice to just be able to say `<Metric … nonNegativeRate />` in the JSX, but I'm a little bummed about the repetition here.

I guess changing all the metrics to `<Metric … derivative={NON_NEGATIVE_RATE} />` or whatever would be a lot of work, but might be worth it.

Anyway, maybe at least add a little comment here, explaining the overlapping fields?
</blockquote></details>

Added some information to the comment on this class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070)*
<!-- Sent from Reviewable.io -->

vilterp commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thx for the comment & the layout tweak.

---

Reviewed 8 of 8 files at r3.
Review status: 1 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23070#-:-L6ODnXHN9QRYDVKrOt2:b31vv7h)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Cool stuff, Matt!

On Tue, Feb 27, 2018 at 4:57 PM Matt Tracy <notifications@github.com> wrote:

> Merged #23070 <https://github.com/cockroachdb/cockroach/pull/23070>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/23070#event-1495457172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135HsVwoT2tVVTnWUk6YOlsfFtV7WRks5tZHpigaJpZM4SSl7R>
> .
>
-- 

-- Tobias

vilterp commented:
🎉 next stop, Grafana :P
